### PR TITLE
5pm 3 add aggregate statistics page

### LINKED ISSUE #139: Create Aggregate Statistics page
# User Story

As a user I can find a useful summary of department statistics, in a sortable table format, so that I can easily compare useful statistics from all departments.

# Acceptance Criteria

- [x] In the Statistics dropdown menu, there should be an entry that links to an "aggregate statistics" page
- [x] There is a table on that page which displays a row for each department, and columns containing a subset of the statistics from the four other statistics tabs.
- [x] The table is sortable by any column, and a user can click on each column to sort the table by its values. Feel free to remove this from the issue and make it a separate issue to be completed after this one.

# Implementation Todos

- [x] Choose which statistics are most useful.

Staff suggestion: Create a form where users select a quarter or set of quarters. Clicking "Submit" on the form shows users a table where each department has a row, and columns include # of classes, # of full classes, class occupancy (%), and average class size for that department.

Front end:

- [x] Create a form where users can select a quarter or set of quarters.
- [x] Add a service that allows the page to fetch the chosen department information and set the department JSON in order for it to display.

Back end:

- [x] Add a endpoint to the StatisticsController file that will service fetch requests from the frontend.
- [x] In the StatisticsController file, use MongoDB and aggregate the quarter/department data and return as JSON.

Testing:

- [x] Front end tests pass and there is adequate coverage
- [x] Back end tests pass and there is adequate coverage

### PR DESCRIPTION
**Pull request by J-Miller1**
This pull request closes #139 . Users can select a quarter range and see an aggregation of statistics for each department including number of courses offered, course occupancy, and average class size.

The backend creates an aggregation pipeline to gather the relevant statistics. The frontend displays this information in a table that can be sorted by any of the columns by clicking on that column's header. 

Currently, if the selected quarter interval is too large (greater than or equal to 4 years) the MongoDB query takes too much memory and fails. Therefore, we have implemented a check in the frontend that will only call the back end if the selected quarter interval is less than 4 years.

Appropriate tests have also been introduced.

### CODE REVIEW

**Review by MattReddick — DISMISSED** [review]
mr - approved to merge

**Review by dougyuan — APPROVED** [review]
LGTM!

Page functions well and displays the proper useful statistics.

**Review by pconrad — CHANGES_REQUESTED** [review]
One possible regression that needs to be addressed.

Other than that, LGTM.   Fix the one missing menu item, and I think you are good to go.

Someone on the staff should  do a re-review (along with deploying to qa on heroku and doing exploratory testing) before final approval and merging though.

Nice work.

File: javascript/src/main/components/Nav/AppNavbar.js [hunk 1]
```diff
@@ -8,73 +8,73 @@ import useSWR from "swr";
...
-          )}
-          <LinkContainer to={"/about"}>
+            <NavDropdown.Item href="/history/instructor">Search By Instructor</NavDropdown.Item>
+        </NavDropdown>
+        { isAdmin &&
+          (<LinkContainer to={"/admin"}>
+            <Nav.Link>Admin</Nav.Link>
+          </LinkContainer>)
+        }
+        <LinkContainer to={"/about"}>
```
[the remark this answers is not in the document]
Comment (WadeVaresio) [reply javascript/src/main/components/Nav/AppNavbar.js]: Hi Professor Conrad,

Thank you for bringing this to our attention. I believe this was the result of an error when resolving merge conflicts. I have pushed changes that I believe resolves this issue.

File: javascript/src/main/components/Nav/AppNavbar.js [hunk 2]
```diff
@@ -8,73 +8,73 @@ import useSWR from "swr";
...
-          </NavDropdown>
-          {isAdmin && (
-            <NavDropdown title="Admin">
-              <NavDropdown.Item as={Link} to={"/admin/panel"}>
-                Admin Panel
-              </NavDropdown.Item>
-              <NavDropdown.Item as={Link} to={"/admin/settings"}>
-                Admin Settings
-              </NavDropdown.Item>
-            </NavDropdown>
```
[the remark this answers is not in the document]
Comment (WadeVaresio) [reply javascript/src/main/components/Nav/AppNavbar.js:45]: Above comment and commit 072511c addresses this issue.

File: javascript/src/main/components/Nav/AppNavbar.js [hunk 3]
```diff
@@ -8,73 +8,78 @@ import useSWR from "swr";
...
+          </NavDropdown.Item>
+          <NavDropdown.Item as={Link} to="/statistics/courseOccupancyByDivision">
+            Course Occupancy by Class Division
+          </NavDropdown.Item>
+          <NavDropdown.Item as={Link} to="/statistics/classSize">
+            Average Class Size by Department
+          </NavDropdown.Item>
+          <NavDropdown.Item as={Link} to="/statistics/fullDeptSummary">
+            Full Classes Summary
+          </NavDropdown.Item>
```
Comment (dougyuan) [inline javascript/src/main/components/Nav/AppNavbar.js:63]: Weird... idk how this got here, but this is the link to the page Matt and I started work on. It's just a blank page atm with no functionality. It probably needs to go.

**Review by dougyuan — CHANGES_REQUESTED** [review]
The previous issue is fixed. 

There is an extraneous item in the Statistics nav dropdown that should be removed. Besides that, LGTM!

File: javascript/src/main/components/Nav/AppNavbar.js (hunk 3 above)
Comment (WadeVaresio) [reply javascript/src/main/components/Nav/AppNavbar.js]: Commit 7dd4226 addresses this.

**Review by dougyuan — APPROVED** [review]
LGTM!

File: javascript/src/test/components/Statistics/AggregateStatisticsForm.test.js [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    });
+
+    test("when I select a end quarter year, the state for end quarter year changes", () => {
+        const { getByTestId } = render(<AggregateStatisticsForm />);
+        const selectQuarter = getByTestId("select-end-year")
+        userEvent.selectOptions(selectQuarter, "2019");
+        expect(selectQuarter.value).toBe("2019");
+    });
+
+    test("when I click submit with a good date range, the right stuff happens", async () => {
```
Comment (JacquelineMai) [inline javascript/src/test/components/Statistics/AggregateStatisticsForm.test.js:42]: It'd be good to have more informative names for this and the following test- what is the right stuff that we should expect to happen?

**Review by JacquelineMai — CHANGES_REQUESTED** [review]

**Review by MattReddick — APPROVED** [review]
mr- merge approved

**Review by JacquelineMai — APPROVED** [review]

**Pull request merged**